src/relay/ir/text_printer.cc

```diff
@@ -0,0 +1,752 @@
+    stream_ << suffix;
+  }
+  /*!
+   * \brief Allocate temporary value
+   * \return A new text value.
+   */
+  TextValue AllocTempVar() {
+    std::ostringstream os;
+    os << '%' << temp_var_counter_++;
+    return TextValue(os.str());
+  }
+  /*!
+   * \brief Allocate name to a variable.
+   * \param var The input variable.
+   * \return The corresponding name.
+   */
+  TextValue AllocVarName(const Var& var) {
+    std::string name = GetUniqueName('%' + var->name_hint);
+    TextValue val(name);
+    CHECK(!memo_.count(var));
+    memo_[var] = val;
+    return val;
+  }
+
+ private:
+  class AttrPrinter;
+  friend class AttrPrinter;
+  /*! \brief meta data context */
+  TextMetaDataContext meta_;
+  /*! \brief Check whether scope is still valid */
+  std::vector<bool> scope_valid_;
+  /*! \brief The current indentation value */
+  int indent_{0};
+  /*! \brief name allocation map */
+  std::unordered_map<std::string, int> name_alloc_map_;
+  /*! \brief Map from expression to its text value */
+  std::unordered_map<Expr, TextValue, NodeHash, NodeEqual> memo_;
+  /*! \brief counter of temporary variable */
+  int64_t temp_var_counter_{0};
+  /*! \brief Output stream */
+  std::ostringstream stream_;
+};
+
+/*!
+ * \brief Attribute printer which prints the attributes in the call.
+ */
+class TextPrinter::AttrPrinter: public AttrVisitor {
+ public:
+  AttrPrinter(std::ostream& stream, TextPrinter* parent)  // NOLINT(*)
+      : stream_(stream), parent_(parent) {}
+
+  void Visit(const char* key, double* value) final {
+    PrintSep();
+    stream_ << key << "=" << value[0];
+  }
+  void Visit(const char* key, int64_t* value) final {
+    PrintSep();
+    stream_ << key << "=" << value[0];
+  }
+  void Visit(const char* key, uint64_t* value) final {
+    PrintSep();
+    stream_ << key << "=" << value[0];
+  }
+  void Visit(const char* key, int* value) final {
+    PrintSep();
+    stream_ << key << "=" << value[0];
+  }
+  void Visit(const char* key, bool* value) final {
+    PrintSep();
+    stream_ << key << "=";
+    parent_->PrintBool(value[0], stream_);
+  }
+  void Visit(const char* key, std::string* value) final {
+    PrintSep();
+    stream_ << key << "=";
+    parent_->PrintString(value[0], stream_);
+  }
+  void Visit(const char* key, void** value) final {
+    LOG(FATAL) << "do not allow void as argument";
+  }
+  void Visit(const char* key, DataType* value) final {
+    PrintSep();
+    stream_ << key << "=";
+    parent_->PrintString(runtime::TVMType2String(Type2TVMType(value[0])), stream_);
+  }
+  void Visit(const char* key, NodeRef* value) final {
+    PrintSep();
+    stream_ << key << "=";
+    parent_->PrintAttr(value[0], stream_);
+  }
+  void Visit(const char* key, runtime::NDArray* value) final {
+    LOG(FATAL) << "do not allow NDarray as argument";
+  }
+
+ private:
+  void PrintSep() {
+    if (counter_ != 0) {
+      stream_ << ", ";
+    }
+    ++counter_;
+  }
+  std::ostream& stream_;  // NOLINT(*)
```

Comment:
Is it safe to assume the ostream is open?

Reply:
yes, it is open during construction time